File: Front-End/src/services/dashboard.service.ts
import axiosClient from '@/configurations/axios.config';
import type {
  RevenueByMonthResponse,
  RevenueByDayResponse,
  RevenueByYearResponse,
  TopProductResponse,
  DashboardApiResponse,
  ComparisonResponse,
  DashboardStatsResponse
} from '@/types/dashboard.type';

export const dashboardService = {
  // Doanh thu theo tháng
  getRevenueByMonth: async (year?: number, month?: number) => {
    const params = new URLSearchParams();
    if (year) params.append('year', year.toString());
    if (month) params.append('month', month.toString());
    
    const queryString = params.toString() ? `?${params.toString()}` : '';
    const response = await axiosClient.get<DashboardApiResponse<RevenueByMonthResponse[]>>(
      `/dashboard/revenue-by-month${queryString}`
    );
    return response.data;
  },

  // Doanh thu theo ngày
  getRevenueByDay: async (startDate?: string, endDate?: string) => {
    const params = new URLSearchParams();
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);
    
    const queryString = params.toString() ? `?${params.toString()}` : '';
    const response = await axiosClient.get<DashboardApiResponse<RevenueByDayResponse[]>>(
      `/dashboard/revenue-by-day${queryString}`
    );
    return response.data;
  },

  // Doanh thu theo năm
  getRevenueByYear: async (year?: number) => {
    const params = year ? `?year=${year}` : '';
    const response = await axiosClient.get<DashboardApiResponse<RevenueByYearResponse[]>>(
      `/dashboard/revenue-by-year${params}`
    );
    return response.data;
  },

  // Top sản phẩm theo ngày
  getTopProductsByDay: async (startDate?: string, endDate?: string) => {
    const params = new URLSearchParams();
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);
    
    const queryString = params.toString() ? `?${params.toString()}` : '';
    const response = await axiosClient.get<DashboardApiResponse<TopProductResponse[]>>(
      `/dashboard/top-products-by-day${queryString}`
    );
    return response.data;
  },

  // Top sản phẩm theo tháng
  getTopProductsByMonth: async (year?: number, month?: number) => {
    const params = new URLSearchParams();
    if (year) params.append('year', year.toString());
    if (month) params.append('month', month.toString());
    
    const queryString = params.toString() ? `?${params.toString()}` : '';
    const response = await axiosClient.get<DashboardApiResponse<TopProductResponse[]>>(
      `/dashboard/top-products-by-month${queryString}`
    );
    return response.data;
  },

  // Top sản phẩm theo năm
  getTopProductsByYear: async (year?: number) => {
    const params = year ? `?year=${year}` : '';
    const response = await axiosClient.get<DashboardApiResponse<TopProductResponse[]>>(
      `/dashboard/top-products-by-year${params}`
    );
    return response.data;
  },

  // So sánh doanh thu
  compareRevenue: async (
    timeType: string,
    startDate1: string,
    endDate1: string,
    startDate2: string,
    endDate2: string
  ) => {
    const params = new URLSearchParams();
    params.append('timeType', timeType);
    params.append('startDate1', startDate1);
    params.append('endDate1', endDate1);
    params.append('startDate2', startDate2);
    params.append('endDate2', endDate2);
    
    const response = await axiosClient.get<DashboardApiResponse<ComparisonResponse>>(
      `/dashboard/compare-revenue?${params.toString()}`
    );
    return response.data;
  },

  // Thống kê tổng quan
  getDashboardStats: async (startDate?: string, endDate?: string) => {
    const params = new URLSearchParams();
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);
    
    const queryString = params.toString() ? `?${params.toString()}` : '';
    const response = await axiosClient.get<DashboardApiResponse<DashboardStatsResponse>>(
      `/dashboard/stats${queryString}`
    );
    return response.data;
  },

  // Export Excel
  exportExcel: async (startDate?: string, endDate?: string): Promise<Blob> => {
    const params = new URLSearchParams();
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);
    
    const queryString = params.toString() ? `?${params.toString()}` : '';
    const response = await axiosClient.get(`/dashboard/export-excel${queryString}`, {
      responseType: 'blob',
    });
    return response.data;
  },

  // Get all products by day
  getAllProductsByDay: async (startDate?: string, endDate?: string) => {
    const params = new URLSearchParams();
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);
    
    const queryString = params.toString() ? `?${params.toString()}` : '';
    const response = await axiosClient.get<DashboardApiResponse<TopProductResponse[]>>(
      `/dashboard/all-products-by-day${queryString}`
    );
    return response.data;
  },

  // Statistics API - Dashboard overview
  getStatisticsDashboard: async () => {
    const response = await axiosClient.get<DashboardApiResponse<any>>(
      '/statistics/dashboard'
    );
    return response.data;
  },

  // Get all products by month
  getAllProductsByMonth: async (year?: number, month?: number) => {
    const params = new URLSearchParams();
    if (year) params.append('year', year.toString());
    if (month) params.append('month', month.toString());
    
    const queryString = params.toString() ? `?${params.toString()}` : '';
    const response = await axiosClient.get<DashboardApiResponse<TopProductResponse[]>>(
      `/dashboard/all-products-by-month${queryString}`
    );
    return response.data;
  },

  // Get all products by year
  getAllProductsByYear: async (year?: number) => {
    const params = year ? `?year=${year}` : '';
    const response = await axiosClient.get<DashboardApiResponse<TopProductResponse[]>>(
      `/dashboard/all-products-by-year${params}`
    );
    return response.data;
  },

  // Get product detail
  getProductDetail: async (productId: number, startDate?: string, endDate?: string) => {
    const params = new URLSearchParams();
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);
    
    const queryString = params.toString() ? `?${params.toString()}` : '';
    const response = await axiosClient.get<DashboardApiResponse<import('@/types/dashboard.type').ProductDetailResponse>>(
      `/dashboard/product-detail/${productId}${queryString}`
    );
    return response.data;
  },

  // Get orders by date range
  getOrdersByDateRange: async (startDate?: string, endDate?: string) => {
    const params = new URLSearchParams();
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);
    
    const queryString = params.toString() ? `?${params.toString()}` : '';
    const response = await axiosClient.get<DashboardApiResponse<import('@/types/dashboard.type').OrderSummary[]>>(
      `/dashboard/orders-by-date-range${queryString}`
    );
    return response.data;
  },

  // Statistics API - Monthly revenue
  getStatisticsMonthlyRevenue: async () => {
    const response = await axiosClient.get<DashboardApiResponse<any[]>>(
      '/statistics/monthly-revenue'
    );
    return response.data;
  }
};

File: Back-End/src/main/java/iuh/fit/ecommerce/services/DashboardService.java
package iuh.fit.ecommerce.services;

import iuh.fit.ecommerce.dtos.response.dashboard.*;

import java.time.LocalDate;
import java.util.List;

public interface DashboardService {
    List<RevenueByMonthResponse> getRevenueByMonth(Integer year, Integer month);
    List<RevenueByDayResponse> getRevenueByDay(LocalDate startDate, LocalDate endDate);
    List<RevenueByYearResponse> getRevenueByYear(Integer year);
    List<TopProductResponse> getTopProductsByDay(LocalDate startDate, LocalDate endDate);
    List<TopProductResponse> getTopProductsByMonth(Integer year, Integer month);
    List<TopProductResponse> getTopProductsByYear(Integer year);
    
    // So sánh doanh thu
    ComparisonResponse compareRevenue(
        String timeType,
        LocalDate startDate1, LocalDate endDate1,
        LocalDate startDate2, LocalDate endDate2
    );
    
    // Top 5 vouchers
    List<TopVoucherResponse> getTopVouchersByDay(LocalDate startDate, LocalDate endDate);
    List<TopVoucherResponse> getTopVouchersByMonth(Integer year, Integer month);
    List<TopVoucherResponse> getTopVouchersByYear(Integer year);
    
    // Top 5 promotions
    List<TopPromotionResponse> getTopPromotionsByDay(LocalDate startDate, LocalDate endDate);
    List<TopPromotionResponse> getTopPromotionsByMonth(Integer year, Integer month);
    List<TopPromotionResponse> getTopPromotionsByYear(Integer year);
    
    // So sánh voucher
    VoucherComparisonResponse compareVoucher(
        String timeType,
        LocalDate startDate1, LocalDate endDate1,
        LocalDate startDate2, LocalDate endDate2
    );
    
    // So sánh promotion
    PromotionComparisonResponse comparePromotion(
        String timeType,
        LocalDate startDate1, LocalDate endDate1,
        LocalDate startDate2, LocalDate endDate2
    );
    
    // So sánh voucher vs promotion (tổng hợp)
    VoucherPromotionComparisonResponse compareVoucherPromotion(
        String timeType,
        LocalDate startDate1, LocalDate endDate1,
        LocalDate startDate2, LocalDate endDate2
    );
    
    // Thống kê tổng quan dashboard
    DashboardStatsResponse getDashboardStats(LocalDate startDate, LocalDate endDate);
    
    // Chi tiết voucher với danh sách đơn hàng
    VoucherDetailResponse getVoucherDetail(Long voucherId, LocalDate startDate, LocalDate endDate);
    
    // ALL vouchers (không giới hạn top 5)
    List<TopVoucherResponse> getAllVouchersByDay(LocalDate startDate, LocalDate endDate);
    List<TopVoucherResponse> getAllVouchersByMonth(Integer year, Integer month);
    List<TopVoucherResponse> getAllVouchersByYear(Integer year);
    
    // ALL promotions (không giới hạn top 5)
    List<TopPromotionResponse> getAllPromotionsByDay(LocalDate startDate, LocalDate endDate);
    List<TopPromotionResponse> getAllPromotionsByMonth(Integer year, Integer month);
    List<TopPromotionResponse> getAllPromotionsByYear(Integer year);
    
    // Chi tiết promotion với danh sách đơn hàng
    PromotionDetailResponse getPromotionDetail(Long promotionId, LocalDate startDate, LocalDate endDate);
    
    // ALL products (không giới hạn top 5)
    List<TopProductResponse> getAllProductsByDay(LocalDate startDate, LocalDate endDate);
    List<TopProductResponse> getAllProductsByMonth(Integer year, Integer month);
    List<TopProductResponse> getAllProductsByYear(Integer year);
    
    // Chi tiết sản phẩm với danh sách đơn hàng
    ProductDetailResponse getProductDetail(Long productId, LocalDate startDate, LocalDate endDate);
    
    // Danh sách đơn hàng theo khoảng thời gian
    List<OrderSummaryResponse> getOrdersByDateRange(LocalDate startDate, LocalDate endDate);
}
